### ISSUE TITLE
Terraform 0.9.8 always wanting to change aws_lambda_event_source_mapping

### ISSUE THREAD

**Issue by joshmyers** [terraform-providers/terraform-provider-aws#929]
Terraform 0.9.8 (but I think a regression was released into 0.9.7, see PR in References) wants to continuously change `aws_lambda_event_source_mapping` on every Terraform run.

### Terraform Version
Terraform 0.9.8

### Affected Resource(s)
- aws_lambda_event_source_mapping

### Terraform Configuration Files
```hcl
resource "aws_lambda_event_source_mapping" "logs" {
  count             = "${length(keys(var.logs))}"
  batch_size        = "${var.lambda_event_mapping_batch_size}"
  event_source_arn  = "${element(aws_kinesis_stream.logs.*.arn, count.index)}"
  enabled           = true
  function_name     = "${var.environment}__${element(keys(var.logs), count.index)}"
  starting_position = "TRIM_HORIZON"
}

resource "aws_kinesis_stream" "logs" {
  count       = "${length(keys(var.logs))}"
  name        = "${var.environment}__${element(keys(var.logs), count.index)}"
  shard_count = "${lookup(var.logs, element(keys(var.logs), count.index))}"

  # FIXME - https://github.com/hashicorp/terraform/issues/11816
  lifecycle {
    ignore_changes = ["shard_count"]
  }

  tags {
    Environment = "${var.environment}"
  }
}
```

### Debug Output
https://gist.github.com/joshmyers/f3e6f901e8ddd2a5f8ba0e8eae8556a8 is just the output of changes Terraform wants to perform on every run, I can add a full debug output if needed.

### Expected Behavior
Expected behaviour as of Terraform 0.9.6:
terraform plan + apply + plan sees no changes in the last plan.

### Actual Behavior
Observed Behaviour as of Terraform 0.9.8:
terraform plan + apply + plan wants to change mappings on every plan/apply

### References
- https://github.com/hashicorp/terraform/pull/14898

**Comment by apparentlymart** [comment]
Thanks for reporting this, @joshmyers.

Indeed it looks like the PR you referenced included [setting the function name to the ARN during read](https://github.com/hashicorp/terraform/blob/6fcf589eca977e1487f209999f16af832c2d430d/builtin/providers/aws/resource_aws_lambda_event_source_mapping.go#L157), which doesn't match how this attribute is being used elsewhere.

This is rather awkward because [the underlying API](http://docs.aws.amazon.com/lambda/latest/dg/API_GetEventSourceMapping.html) returns only the ARN here, while the _write_ endpoints for this object accept either a function name _or_ an ARN. To fix this, I expect we will need to parse the ARN to extract the function name in order to correctly refresh the state.

**Comment by bflad** [comment]
Hi @joshmyers 👋 Apologies for missing this issue earlier! It appears that #5454 was merged into the AWS provider that should prevent this issue by suppressing the difference between function name and ARN. This was previously released in [version 1.34.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1340-august-30-2018) and has been available in all releases since. Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

If the mentioned AWS provider version does not fix the issue, I would suggest creating a fresh bug report on recent versions of Terraform and the AWS provider so we can troubleshoot further or provide more current guidance. Thanks!
